MoreLinq/SelectAsync.cs

```diff
@@ -0,0 +1,439 @@
+#region License and Terms
+// MoreLINQ - Extensions to LINQ to Objects
+// Copyright (c) 2016 Atif Aziz. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#endregion
+
+#if !NO_ASYNC
+
+namespace MoreLinq
+{
+    using System;
+    using System.Collections;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using System.Linq;
+    using System.Runtime.ExceptionServices;
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// Represents options for an asynchronous projection operation.
+    /// </summary>
+
+    public sealed class SelectAsyncOptions
+    {
+        /// <summary>
+        /// The default options an asynchronous projection operation.
+        /// </summary>
+
+        public static readonly SelectAsyncOptions Default =
+            new SelectAsyncOptions(null /* = unbounded concurrency */,
+                                   TaskScheduler.Default,
+                                   preserveOrder: false);
+
+        /// <summary>
+        /// Gets a positive (non-zero) integer that specifies the maximum
+        /// projections to run concurrenctly or <c>null</c> to mean unlimited
+        /// concurrency.
+        /// </summary>
+
+        public int? MaxConcurrency { get; }
```

Comment:
Why not `int`? Zero is not allowed anyway.

Reply:
Because it can be undefined or unspecified, which means unbounded concurrency.